README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rweatherlink

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/rweatherlink)](https://CRAN.R-project.org/package=rweatherlink)
[![R build
status](https://github.com/basf/rweatherlink/workflows/R-CMD-check/badge.svg)](https://github.com/basf/rweatherlink/actions)
[![Codecov test
coverage](https://codecov.io/gh/basf/rweatherlink/branch/master/graph/badge.svg?token=3OZ8Y9VVWN)](https://codecov.io/gh/basf/rweatherlink?branch=master)
<!-- badges: end -->

An R client for [Weatherlink v2
API](https://weatherlink.github.io/v2-api/)

## Usage

``` r
library("rweatherlink")
```

### Authentication

The api key and api secret are read by default from environmental
variables `WL_API_KEY` and `WL_API_SECRET`, but you can provide them
also in every function call using the `api_key=` and `api_secret=`
arguments.

### Basic usage

For each endpoint we provide convenience wrappers, like

``` r
wl_stations()
```

The full list of endpoints wrappers is:

  - `wl_stations()` for `/stations` & `stations/{station-ids}` endpoints
  - `wl_nodes()` for `/nodes` & `nodes/{nodes-ids}` endpoints
  - `wl_sensors()` for `/sensors` & `sensors/{sensor-ids}` endpoints
  - `wl_sensor_activity()` for `/sensor-activity` &
    `sensor-activity/{sensor-ids}` endpoints
  - `wl_sensor_catalog()` for `/sensor-catalog` endpoint
  - `wl_current()` for `/current/{station-id}` endpoint
  - `wl_historic()` for `/historic/{station-id}` endpoint

### Parser

For each of the listed functions (except `wl_sensor_catalog()`) we
provide (opinionated) convenience parsers, that parse the return value
into a nice `data.frame`. Columns with UTC-dates for timestamps are
added, as well as elevation in meters.

The generic function for all objects is `wl_parse()`

``` r
wl_stations() %>% wl_parse()
wl_sensors() %>% wl_parse()
```

These parsers are in an experimental stage.

### Timestamps

`wl_historic()` needs unix-timestamps as input for start and end. These
can be easily generated from `POSIXct` objects, e.g.

``` r
as.integer(Sys.time())
#> [1] 1608050438
```

gives the timestamp of the current time.

## Under the hood

`wl_request()` is the workhorse of this package.

With it you can query every API endpoint, e.g.

``` r
wl_request(endpoint = "stations")
```

It takes also care of creating the correct api-signature via HMAC, which
us handled by `wl_params()`.

## Unit tests

To run the full test suite set valid environmental variables
`WL_API_KEY` and `WL_API_SECRET` and run `devtools::test()`.

Note, that for some tests more than one stations is needed.

Currently, code coverage is very high. wl\_nodes is not well covered by
the unit tests.


NEWS.md
# rweatherlink 0.1.0

* Initial release.


tests/testthat.R
library(testthat)
library(rweatherlink)

test_check("rweatherlink")


tests/testthat/test-parsers.R
context("parsers")
skip_if(Sys.getenv("WL_API_KEY") == "" | Sys.getenv("WL_API_SECRET") == "")

stations <- wl_stations()
skip_if(length(stations$stations) < 1)


# wl_stations ----------------------------------------------------------------
test_that("correctly parse stations w/o agrs", {
  res <- stations %>% wl_parse()
  expect_is(res, "data.frame")
  expect_true(ncol(res) == 24)
  expect_is(res$registered_date_utc, "POSIXct")
})
test_that("correctly parse stations one arg", {
  res <- wl_stations(station_ids = stations$stations[[1]]$station_id) %>%
    wl_parse()
  expect_is(res, "data.frame")
  expect_true(ncol(res) == 24)
  expect_true(nrow(res) == 1)
  expect_is(res$registered_date_utc, "POSIXct")
  expect_equal(attr(res$registered_date_utc, "tzone"), "UTC")
})
test_that("correctly parse stations two args", {
  skip_if(length(stations$stations) < 2)
  res <- wl_stations(station_ids = c(stations$stations[[1]]$station_id,
                                     stations$stations[[2]]$station_id)) %>%
    wl_parse()
  expect_is(res, "data.frame")
  expect_true(ncol(res) == 24)
  expect_true(nrow(res) == 2)
  expect_is(res$registered_date_utc, "POSIXct")
  expect_equal(attr(res$registered_date_utc, "tzone"), "UTC")
})


# wl_nodes -------------------------------------------------------------------
nodes <- wl_nodes()
test_that("correctly parse nodes w/o agrs", {
  res <- nodes %>% wl_parse()
  expect_equal(res, NULL)
})


# wl_sensors -----------------------------------------------------------------
sensors <- wl_sensors()
test_that("correctly parse sensors w/o agrs", {
  res <- sensors %>% wl_parse()
  expect_is(res, "data.frame")
  expect_true(ncol(res) == 24)
  expect_is(res$created_date_utc, "POSIXct")
})
test_that("correctly parse sensors one arg", {
  res <- wl_sensors(sensor_ids = sensors$sensors[[1]]$lsid) %>%
    wl_parse()
  expect_is(res, "data.frame")
  expect_true(nrow(res) == 1)
  expect_true(ncol(res) == 24)
  expect_is(res$created_date_utc, "POSIXct")
  expect_equal(attr(res$created_date_utc, "tzone"), "UTC")
})
test_that("correctly parse sensors with 2 args", {
  skip_if(length(sensors$sensors) < 2)
  res <- wl_sensors(sensor_ids = c(sensors$sensors[[1]]$lsid,
                                   sensors$sensors[[2]]$lsid)) %>%
    wl_parse()
  expect_is(res, "data.frame")
  expect_true(nrow(res) == 2)
  expect_true(ncol(res) == 24)
  expect_is(res$created_date_utc, "POSIXct")
  expect_equal(attr(res$created_date_utc, "tzone"), "UTC")
})


# wl_sensor_activity ---------------------------------------------------------
activity <- wl_sensor_activity()
test_that("correctly parse sensor_activity w/o agrs", {
  expect_warning({
    res <- activity %>% wl_parse()
    },
    regexp = "No matching parsing method for object of class")
  expect_is(res, "wl_sensor_activity")
  expect_equal(res, activity)
})

# wl_sensor_activity ---------------------------------------------------------
catalog <- wl_sensor_catalog()
test_that("correctly parse sensor_activity w/o agrs", {
  expect_warning({
    res <- catalog %>% wl_parse()
  },
  regexp = "No matching parsing method for object of class")
  expect_is(res, "wl_sensor_catalog")
  expect_equal(res, catalog)
})


# wl_current --------------------------------------------------------------
current <- wl_current(station_id = stations$stations[[1]]$station_id)
test_that("correctly parse current w/o agrs", {
  res <- current %>% wl_parse()
  expect_is(res, "data.frame")
  expect_true(nrow(res) == 1)
  expect_is(res$ts_utc, "POSIXct")
  expect_equal(attr(res$ts_utc, "tzone"), "UTC")
})


# wl_historic -------------------------------------------------------------

historic <- wl_historic(station_id = stations$stations[[1]]$station_id,
                   start = as.integer(as.POSIXct("2020-06-16 13:59:39 CET")),
                   end = as.integer(as.POSIXct("2020-06-16 20:59:39 CET")))
test_that("correctly parse historic w/o agrs", {
  res <- historic %>% wl_parse()
  expect_is(res, "data.frame")
  expect_is(res$ts_utc, "POSIXct")
  expect_equal(attr(res$ts_utc, "tzone"), "UTC")
})


tests/testthat/test-request.R
context("wl_request")

test_that("wl_request fails with invalid credentials", {
  expect_error(
    wl_request(endpoint = "stations",
               api_key = "invalid",
               api_secret = "invalid"),
    regexp = "Invalid authentication credentials")
})

test_that("wl_request fails with missing credentials", {
  skip_if(Sys.getenv("WL_API_KEY") == "" | Sys.getenv("WL_API_SECRET") == "")

  expect_error(wl_request(endpoint = "stations", api_key = ""),
               regexp = "api_key is empty")
  expect_error(wl_request(endpoint = "stations", api_secret = ""),
               regexp = "api_secret is empty")
})

test_that("wl_request works", {
  skip_if(Sys.getenv("WL_API_KEY") == "" | Sys.getenv("WL_API_SECRET") == "")

  expect_message({
    stations <- wl_request(endpoint = "stations", verbose = TRUE)
    },
    "GET: https://api.weatherlink.com/v2/stations")
  expect_is(stations, "list")
  expect_true(names(stations)[[1]] == "stations")
})


R/parsers.R
#' parse output from rweatherlink objects
#' @param x weatherlink object
#' @param ... other arguments to methods (currently unused)
#' @export
#' @examples
#' \dontrun{
#' (stations <- wl_stations() %>% wl_parse())
#' wl_nodes() %>% wl_parse()
#' wl_sensors() %>% wl_parse()
#' wl_sensor_activity() %>% wl_parse()
#' wl_current(station_id = stations$station_id[[1]]) %>% wl_parse()
#' wl_historic(station_id = stations$station_id[[1]],
#'   start = as.integer(as.POSIXct("2020-06-16 13:59:39 CET")),
#'   end = as.integer(as.POSIXct("2020-06-16 20:59:39 CET"))
#'  ) %>% wl_parse()
#' }
wl_parse <- function(x, ...) {
  UseMethod("wl_parse", object = x)
}

#' @export
wl_parse.default <- function(x, ...) {
  warning("No matching parsing method for object of class '",
          class(x)[[1]], "' found. Returning object."
  )
  x
}

#' @export
#' @importFrom data.table setDF rbindlist
wl_parse.wl_stations <- function(x, ...) {
  if (length(x$stations) == 0)
    stop("No stations to parse.")

  suppressWarnings({
    out <- data.table::setDF(data.table::rbindlist(x$stations, fill = TRUE))
  })

  if (nrow(out) == 0)
    return(NULL)

  out <- out %>%
    mutate(elevation_m = .data$elevation / 3.2808,
           registered_date_utc = as.POSIXct(.data$registered_date,
                                            origin = "1970-01-01",  tz = "UTC"))
  return(out)
}

#' @export
#' @importFrom data.table setDF rbindlist
wl_parse.wl_nodes <- function(x, ...) {
  if (length(x$nodes) == 0)
    stop("No nodes to parse.")
  suppressWarnings({
    out <- data.table::setDF(data.table::rbindlist(x$nodes, fill = TRUE))
  })
  if (nrow(out) == 0)
    return(NULL)
  return(out)
}

#' @export
#' @importFrom data.table setDF rbindlist
wl_parse.wl_sensors <- function(x, ...) {
  if (length(x$sensors) == 0)
    stop("No sensors to parse.")
  suppressWarnings({
    out <- data.table::setDF(data.table::rbindlist(x$sensors, fill = TRUE))
  })
  if (nrow(out) == 0)
    return(NULL)

  out <- out %>%
    mutate(elevation_m = .data$elevation / 3.2808,
           created_date_utc = as.POSIXct(.data$created_date,
                                         origin = "1970-01-01",  tz = "UTC"),
           modified_date_utc = as.POSIXct(.data$modified_date,
                                          origin = "1970-01-01",  tz = "UTC"))

  return(out)
}

#' @importFrom data.table setDF rbindlist
parse_sensor <- function(sensor) {
  if (length(sensor$data) == 0)
    stop("No data to parse.")
  suppressWarnings({
    data <- data.table::setDF(data.table::rbindlist(sensor$data, fill = TRUE))
  })
  out <- data.frame(lsid = sensor$lsid, sensor_type = sensor$sensor_type, data)
  return(out)
}

#' @export
#' @importFrom purrr map_df
#' @import dplyr
wl_parse.wl_current <- function(x, ...) {
  out <- purrr::map_df(x$sensors, parse_sensor)
  out <- out %>%
    mutate(ts_utc = as.POSIXct(.data$ts, origin = "1970-01-01",  tz = "UTC"))
  return(out)
}

#' @export
#' @importFrom purrr map_df
#' @import dplyr
wl_parse.wl_historic <- function(x, ...) {
  out <- purrr::map_df(x$sensors, parse_sensor)
  out <- out %>%
    mutate(ts_utc = as.POSIXct(.data$ts, origin = "1970-01-01",  tz = "UTC"))
  return(out)
}


R/wrappers.R
#' Query /stations & stations/{station-ids} endpoint
#' @rdname wl_wrappers
#' @param station_ids vector of stations ids to query. If NULL (default) all
#'   will be returned.
#' @param ... other arguments passed to [wl_request()]
#' @references https://weatherlink.github.io/v2-api/api-reference
#' @export
#' @examples
#' \dontrun{
#' (stations <- wl_stations())
#' wl_stations(station_ids = stations$stations[[1]]$station_id)
#' wl_stations(station_ids = c(stations$stations[[1]]$station_id,
#'     stations$stations[[2]]$station_id))
#' }
wl_stations <- function(station_ids = NULL, ...) {

  if (is.null(station_ids)) {
    out <- wl_request(endpoint = "stations", ...)
  } else {
    out <- wl_request(endpoint = "stations",
      path = list(`station-ids` = station_ids), ...)
  }
  class(out) <- append("wl_stations", class(out))
  return(out)
}

#' Query /nodes & nodes/{nodes-ids} endpoint
#' @rdname wl_wrappers
#' @param node_ids vector of stations ids to query. If NULL (default) all will
#'   be returned.
#' @export
#' @examples
#' \dontrun{
#' wl_nodes()
#' }
wl_nodes <- function(node_ids = NULL, ...) {

  if (is.null(node_ids)) {
    out <- wl_request(endpoint = "nodes", ...)
  } else {
    out <- wl_request(endpoint = "nodes",
                      path = list(`node-ids` = node_ids), ...)
  }
  class(out) <- append("wl_nodes", class(out))
  return(out)
}

#' Query /sensors & sensors/{sensor-ids} endpoint
#' @rdname wl_wrappers
#' @param sensor_ids vector of stations ids to query. If NULL (default) all
#'   will be returned.
#' @export
#' @examples
#' \dontrun{
#' (sensors <- wl_sensors())
#' wl_sensors(sensor_ids = sensors$sensors[[1]]$lsid)
#' wl_sensors(sensor_ids = c(sensors$sensors[[1]]$lsid,
#'   sensors$sensors[[2]]$lsid))
#' }
wl_sensors <- function(sensor_ids = NULL, ...) {

  if (is.null(sensor_ids)) {
    out <- wl_request(endpoint = "sensors", ...)
  } else {
    out <- wl_request(endpoint = "sensors",
                      path = list(`sensor-ids` = sensor_ids), ...)
  }
  class(out) <- append("wl_sensors", class(out))
  return(out)
}

#' Query /sensor-activity & sensor-activity/{sensor-ids} endpoint
#' @rdname wl_wrappers
#' @export
#' @examples
#' \dontrun{
#' wl_sensor_activity()
#' wl_sensor_activity(sensor_ids = sensors$sensors[[1]]$lsid)
#' wl_sensor_activity(sensor_ids = c(sensors$sensors[[1]]$lsid,
#'   sensors$sensors[[2]]$lsid))
#' }
wl_sensor_activity <- function(sensor_ids = NULL, ...) {

  if (is.null(sensor_ids)) {
    out <- wl_request(endpoint = "sensor-activity", ...)
  } else {
    out <- wl_request(endpoint = "sensor-activity",
                      path = list(`sensor-ids` = sensor_ids), ...)
  }
  class(out) <- append("wl_sensor_activity", class(out))
  return(out)
}

#' Query /sensor-catalog
#' @rdname wl_wrappers
#' @export
#' @examples
#' \dontrun{
#' wl_sensor_catalog()
#' }
wl_sensor_catalog <- function(...) {
  out <- wl_request(endpoint = "sensor-catalog", ...)
  class(out) <- append("wl_sensor_catalog", class(out))
  return(out)
}

#' Query /current/{station-id} endpoint
#' @rdname wl_wrappers
#' @param station_id station_id for which to query data.
#' @export
#' @examples
#' \dontrun{
#' (stations <- wl_stations())
#' wl_current(station_id = stations$stations[[1]]$station_id)
#' }
wl_current <- function(station_id, ...) {

  stopifnot(!missing(station_id))
  stopifnot(length(station_id) == 1)
  out <- wl_request(endpoint = "current",
                    path = list(`station-id` = station_id),
                    ...)
  class(out) <- append("wl_current", class(out))
  return(out)
}


#' Query /historic/{station-id} endpoint
#' @rdname wl_wrappers
#' @param start integer; Unix timestamp. Current timestamp can be generated
#'   with as.integer(Sys.time())
#' @param end integer; Unix timestamp. Current timestamp can be generated
#'   with as.integer(Sys.time())
#' @export
#' @examples
#' \dontrun{
#' (stations <- wl_stations())
#' wl_historic(station_id = stations$stations[[1]]$station_id,
#'   start = as.integer(as.POSIXct("2020-06-16 13:59:39 CET")),
#'   end = as.integer(as.POSIXct("2020-06-16 20:59:39 CET"))
#'  )
#' }
wl_historic <- function(station_id, start, end, ...) {

  stopifnot(!missing(station_id))
  stopifnot(length(station_id) == 1)
  stopifnot(!missing(start))
  stopifnot(is.integer(start))
  stopifnot(!missing(end))
  stopifnot(is.integer(end))

  out <- wl_request(endpoint = "historic",
    path = list(`station-id` = station_id),
    query = list(
      `start-timestamp` = start,
      `end-timestamp` = end),
    ...
  )
  class(out) <- append("wl_historic", class(out))
  return(out)
}


R/request.R
#' General request function for weatherlink api
#' @rdname wl_request
#' @param endpoint api endpoint to query
#' @param path named list; path parameter to pass
#' @param query named list of length 2; Only "start-timestamp", "end_timestamp"
#'   are allowed
#' @param api_key API key. Read by default from env variable `WL_API_KEY`
#' @param api_secret API secret. Read by default from env variable
#'   `WL_API_SECRET`
#' @param ignore_ratelimit ignore wait time to not hit rate limit of 10/second?
#' @param verbose logical, should the request be printed?
#' @description authentication is done via hmac, see [wl_params()].
#' @importFrom httr modify_url GET content http_error status_code
#' @importFrom jsonlite fromJSON
#' @importFrom purrr compact
#' @export
#' @examples
#' \dontrun{
#' stations <- wl_request(endpoint = "stations")
#' wl_request(endpoint = "stations",
#'   path = list(`station-ids` = stations$stations[[1]]$station_id))
#' wl_request(endpoint = "stations",
#'   path = list(`station-ids` = c(stations$stations[[1]]$station_id,
#'     stations$stations[[2]]$station_id)))
#' wl_request(endpoint = "current",
#'   path = list(`station-id` = stations$stations[[1]]$station_id))
#' wl_request(endpoint = "historic",
#'   path = list(`station-id` = stations$stations[[1]]$station_id),
#'   query = list(
#'     `start-timestamp` = as.integer(as.POSIXct("2020-06-16 13:59:39 CET")),
#'     `end-timestamp` = as.integer(as.POSIXct("2020-06-16 20:59:39 CET"))))
#' }
wl_request <- function(endpoint = NULL,
                       path = NULL,
                       query = NULL,
                       api_key = Sys.getenv("WL_API_KEY"),
                       api_secret = Sys.getenv("WL_API_SECRET"),
                       ignore_ratelimit = FALSE,
                       verbose = FALSE) {

  # check inputs
  stopifnot(!is.null(endpoint))
  if (!is.null(path)) {
    stopifnot(is.list(path))
    stopifnot(names(path) %in% path_params)
    stopifnot(length(path) == 1)
    path[[1]] <-  paste(path[[1]], collapse = ",")
  }
  if (!is.null(query)) {
    stopifnot(is.list(query))
    stopifnot(names(query) %in% query_params)
    stopifnot(length(query) == 2)
  }

  if (nchar(api_key) == 0)
    stop("api_key is empty. Is environment variable 'WL_API_KEY' set?")

  if (nchar(api_secret) == 0)
    stop("api_secret is empty. Is environment variable 'WL_API_SECRET' set?")

  fullpath <- purrr::compact(list("v2", endpoint, path[[1]]))
  qurl <- httr::modify_url("https://api.weatherlink.com", path = fullpath)
  params <- wl_params(api_key = api_key, path = path, query = query)


  if (verbose)
    message("GET: ", qurl)

  if (!ignore_ratelimit)
    Sys.sleep(0.1)

  resp <- httr::GET(url = qurl, query = params)

  parsed <- resp %>%
    httr::content(as = "text", encoding = "UTF-8") %>%
    jsonlite::fromJSON(simplifyVector = FALSE)

  if (httr::http_error(resp)) {
    stop(
      sprintf(
        "weatherlink API request failed [%s]\n%s",
        httr::status_code(resp),
        parsed$message
      ),
      call. = FALSE
    )
  }

  return(parsed)
}


#' Create paramesters for weatherlink api calls
#' @rdname wl_request
#' @importFrom digest hmac
#' @export
#' @examples
#' \dontrun{
#' wl_params()
#' wl_params(path = list(`station-ids` = 123456))
#' }
wl_params <- function(path = NULL,
                      query = NULL,
                      api_key = Sys.getenv("WL_API_KEY"),
                      api_secret = Sys.getenv("WL_API_SECRET")) {

  if (is.null(path)) {
    params <- list(
      `api-key` = api_key,
      `t` = as.character(as.integer(Sys.time()))
    )
  } else {
    stopifnot(is.list(path))
    stopifnot(names(path) %in% path_params)
    stopifnot(length(path) == 1)
    if (is.null(query)) {
      params <- list(
        `api-key` = api_key,
        p = path[[1]],
        `t` = as.character(as.integer(Sys.time()))
      )
      names(params)[2] <- names(path)
    } else {
      stopifnot(is.list(query))
      stopifnot(names(query) %in% query_params)
      stopifnot(length(query) == 2)
      params <- list(
        `api-key` = api_key,
        `end-timestamp` = query[["end-timestamp"]],
        `start-timestamp` = query[["start-timestamp"]],
        p = path[[1]],
        `t` = as.character(as.integer(Sys.time()))
      )
      names(params)[4] <- names(path)
    }
  }


  signature <- digest::hmac(
    key = api_secret,
    object = paste(names(params),
                   as.character(params), sep = "", collapse = ""),
    algo = "sha256"
  )

  if (is.null(query)) {
    out <- list(`api-key` = params$`api-key`,
                t = params$t,
                `api-signature` = signature)
  } else {
    out <- list(`api-key` = params$`api-key`,
                t = params$t,
                `start-timestamp` = query[["start-timestamp"]],
                `end-timestamp` = query[["end-timestamp"]],
                `api-signature` = signature)
  }

  return(out)
}

#' list of allowed path parameter names
path_params <- c("station-id", "station-ids", "sensor-id", "sensor-ids",
                 "node-ids", "start-timestamp", "end_timestamp")
#' list of allowed query parameter names
query_params <- c("start-timestamp", "end-timestamp")
